### PR TITLE
UPLOAD-1184: go translation of sender manifest validation

### PR DESCRIPTION
**Pull request by whytheplatypus**
This is a translation of the behavior in https://github.com/CDCgov/data-exchange-upload/tree/main/tus/file-hooks/metadata-verify to a native go module.

Given a sender manifest in a pre-create hook:
- [x] load an appropriate config from 
    - [x] filesystem
    - [x] blobstore
    - [x] establish a pattern for any
- [x] load any given config only once
- [x] validate required fields and allowed values against the found config

example for running locally: `go run ./cmd/main.go -appconf ./configs/local/local.env`

### CODE REVIEW

File: upload-server/internal/metadata/metadata.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	var errs error
+	for _, field := range config.Fields {
+		if err := field.Validate(manifest); err != nil {
+
+			if _, ok := err.(*validation.ErrorMissing); !ok {
+				logger.Error("validation failure", "error", err)
+				errs = errors.Join(errs, err)
+			} else {
+				logger.Info("validation warning", "error", err)
+			}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:105]: Can have errors that define severity: `ErrValidationWarning` `ErrValidationFailure`

File: upload-server/internal/metadata/metadata.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			return resp, nil
+		}
+		if errors.Is(err, os.ErrNotExist) {
+			resp.HTTPResponse = resp.HTTPResponse.MergeWith(handler.HTTPResponse{
+				StatusCode: http.StatusBadRequest,
+				Body:       err.Error(),
+			})
+			resp.RejectUpload = true
+			return resp, nil
+		}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:83]: With error severity we'll be able to collapse this to `errors.Is(err, os.ErrNotExist) || errors.Is(err, ErrValidationFailure)`

File: upload-server/internal/metadata/metadata.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		if err != nil {
+			return nil, err
+		}
+		return getter.GetConfig(loader)
+	case "2.0":
+		getter, err := v2.NewFromManifest(manifest)
+		if err != nil {
+			return nil, err
+		}
+		return getter.GetConfig(loader)
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:51]: move to a map

File: upload-server/internal/metadata/metadata.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	type Empty struct{}
+	pkgParts := strings.Split(reflect.TypeOf(Empty{}).PkgPath(), "/")
+	// add package name to app logger
+	logger = sloger.With("pkg", pkgParts[len(pkgParts)-1])
+}
+
+var RequiredManifestFields = map[string][]string{
+	"1.0": {"meta_destination_id", "meta_ext_event"},
+	"2.0": {"data_stream_id", "data_stream_route"},
+}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:31]:
```suggestion
}
```

File: upload-server/internal/metadata/metadata.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+func (v *SenderManifestVerification) Verify(event handler.HookEvent) (hooks.HookResponse, error) {
+	resp := hooks.HookResponse{}
+
+	manifest := event.Upload.MetaData
+	logger.Info("checking the sender manifest:", "manifest", manifest)
+
+	config, err := getVersionFromManifest(event.Context, manifest, v.Loader)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (whytheplatypus) [reply upload-server/internal/metadata/metadata.go:79]: i think i take this back

File: upload-server/internal/metadata/validation/errors.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	ErrFailure = errors.New("validation failure")
+	ErrWarning = errors.New("validation warning")
+)
+
+type ErrorMissing struct {
+	Field string
+}
+
+func (e *ErrorMissing) Error() string {
+	return fmt.Sprintf("field %s was missing", e.Field)
```
Comment (dancing-goats) [inline upload-server/internal/metadata/validation/errors.go:18]: Instead of fmt, let's use log package.

**Comment by cosmycx** [comment]
I'm not able to compile and run locally. It has errors in main, hooks, and metadatav1.

**Review by cosmycx — COMMENTED** [review]
Not able to compile, it has errors in main, hooks, and metadatav1.

File: upload-server/internal/metadata/validation/errors.go (hunk 6 above)
Comment (whytheplatypus) [reply upload-server/internal/metadata/validation/errors.go:18]: this is for returning a string, fmt.Sprintf is a string constructor. log outputs the string to standard out, which fmt , confusingly, also has a method for: [`fmt.Printf`](https://pkg.go.dev/fmt#Printf)

File: upload-server/internal/metadata/metadata.go [hunk 7]
```diff
@@ -40,7 +40,7 @@ type configCache struct {
 	sync.Map
 }
 
-func (c *configCache) GetConfig(key any) (*validation.MetadataConfig, bool) {
+func (c *configCache) GetConfig(key interface{}) (*validation.MetadataConfig, bool) {
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:43]: `any` is correct here, go is cleaning up this confusing naming

File: upload-server/internal/metadata/metadata.go [hunk 8]
```diff
@@ -72,7 +72,7 @@ func loadConfig(ctx context.Context, path string, loader validation.ConfigLoader
 
 func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, loader validation.ConfigLoader) (*validation.MetadataConfig, error) {
 	version, ok := manifest["version"]
-	if !ok {
+	if version == "" {
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:75]: i like this as a stronger check

**Comment by whytheplatypus** [comment]
> Not able to compile, it has errors in main, hooks, and metadatav1.

@cosmycx fixed, added ci checks for both vetting and build

**Review by cosmycx — APPROVED** [review]

**Review by dancing-goats — APPROVED** [review]

**Pull request merged**